Lib/os.py

```diff
@@ -340,43 +340,35 @@ def walk(top, topdown=True, onerror=None, followlinks=False):
 
     """
     sys.audit("os.walk", top, topdown, onerror, followlinks)
+    top = fspath(top)
+    if topdown:
+        return _walk_topdown(top, onerror, followlinks)
+    return _walk_bottomup(top, onerror, followlinks)
 
-    stack = [fspath(top)]
-    islink, join = path.islink, path.join
-    while stack:
-        top = stack.pop()
-        if isinstance(top, tuple):
-            yield top
-            continue
+def _walk_topdown(top, onerror, followlinks):
+    try:
+        scandir_it = scandir(top)
+    except OSError as error:
+        if onerror is not None:
+            onerror(error)
+        return
 
+    stack = []
+    islink, join = path.islink, path.join
+    while True:
         dirs = []
         nondirs = []
-        walk_dirs = []
-
-        # We may not have read permission for top, in which case we can't
-        # get a list of the files the directory contains.
-        # We suppress the exception here, rather than blow up for a
-        # minor reason when (say) a thousand readable directories are still
-        # left to visit.
-        try:
-            scandir_it = scandir(top)
-        except OSError as error:
-            if onerror is not None:
-                onerror(error)
-            continue
-
-        cont = False
+        use_entry = True
         with scandir_it:
             while True:
                 try:
-                    try:
-                        entry = next(scandir_it)
-                    except StopIteration:
-                        break
+                    entry = next(scandir_it)
+                except StopIteration:
+                    break
                 except OSError as error:
                     if onerror is not None:
                         onerror(error)
-                    cont = True
+                    use_entry = False
                     break
```

Comment:
Note the change from
```python
                try:
                    try:
                        entry = next(scandir_it)
                    except StopIteration:
                        break
                except OSError as error:
```
to
```python
                try:
                    entry = next(scandir_it)
                except StopIteration:
                    break
                except OSError as error:
```
This seemed cleaner and equivalent, unless there was a particular reason for the original form